### ISSUE TITLE
Incorrect line in max-empty-lines

### ISSUE THREAD

**Issue by b8kkyn** [stylelint/stylelint#2343]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Bugs, opinions and one idea.

> Which rule, if any, is this issue related to?

`max-empty-lines`

> What CSS is needed to reproduce this issue?

```css





/*
  The above should be warning because my config is "max-empty-lines": 2 
  And there is 3 empty lines and no warning, how?
  Because max-empty-lines doesn't consider beginning of the file.
*/

.selector {
  color: #fff;
}

/**
 * Call me Ishmael.
 *
 *
 *
 * Some years ago -- never mind how log precisely -- ...
 *
 */
/* 
 * Lets look at this. Rule's documentation says this comment block has empty lines and should be warning.
 * Again there is no warning. But in my opinion, shouldn't be either.
 * Because comment block has not actually empty lines, right?
 * Please pay attention to the stars on the left.
 */
 
#samerule {
  background-image: url('funnybugishere.png');
}

/*






 40th Line: Now we have empty lines in comment block. Following lines are considered warning: 37, 38, 39 and 40
 Which are totaly weird. 40th line is not empty and you can see that by looking previous line.
 And again in my opinion these lines should be warning for this comment block: 36, 37, 38 and 39
 By letting only 2 adjacent empty lines ("max-empty-lines": 2)
*/
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "max-empty-lines": 2
  }
}
```

> Which version of stylelint are you using?

`7.8.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Doesn't matter.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No.

> What did you expect to happen?

I expected this:
```shell
something.css
  3:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
 36:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
 37:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
 38:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
 39:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
```

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```shell
something.css
 37:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
 38:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
 39:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
 40:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
```

> So tell us about your idea?

Well, do you guys consider about adding optional secondary option `ignore: ['comment']` to the rule?
This would be a good improving I think. Comments should be ignored with secondary option for `max-empty-line`.

> Last thing, have you ever dealt with the code side of this?

Yes, I have. In fact, I solved all of these (including ignore option). But I didn't test with stylelint way. Because I don't know how I can.  Believe me, there is a something about documentation. It is my first time with Github and Github things. So I don't know how I can help the development of this project.  What is the procedure? 

Sorry for my language. It is not my native language.

**Comment by evilebottnawi** [comment]
@Leucetios i think it is not bug.
```
 3:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
```
For this be good add option `maxEOF` and `maxBOF` (maybe best names) for controlling empty lines in start and end of file.
Also i agree about `ignore: ['comment']` option.
Other error very strange, i have another lines:
```
  5:1  ✖  Expected no more than 2 empty line(s)   max-empty-lines
  6:1  ✖  Expected no more than 2 empty line(s)   max-empty-lines
 39:1  ✖  Expected no more than 2 empty line(s)   max-empty-lines
 40:1  ✖  Expected no more than 2 empty line(s)   max-empty-lines
 41:1  ✖  Expected no more than 2 empty line(s)   max-empty-lines
 42:1  ✖  Expected no more than 2 empty line(s)   max-empty-lines
```

**Comment by evilebottnawi** [comment]
Also i think rule should report about 4 line.

**Comment by davidtheclark** [comment]
Sure, `ignore: ['comment']` seems fine to me.

And it looks like there is a bug with the line counts within comments.

Let's split the above two items into separate issues.

> For this be good add option maxEOF and maxBOF (maybe best names) for controlling empty lines in start and end of file.

Why not just enforce at the beginning and end of the file? I don't see a need for these additional options. Maybe the problem is that this rule uses `root.toString()` — I wonder if that trims the string.

**Comment by evilebottnawi** [comment]
@davidtheclark i am testing with `root.source.input.css`, error have not correct position, but we should change `root.toString()` to `root.source.input.css` because this may cause further errors

**Comment by jeddy3** [comment]
> It is my first time with Github and Github things. So I don't know how I can help the development of this project. What is the procedure?

You'll find plenty of useful info in this [github guide to contributing](https://guides.github.com/activities/contributing-to-open-source/).

After you've read that have a look at [stylelint's developer guide](https://stylelint.io/developer-guide/rules/).

Good luck!

**Comment by Leucetios** [comment]
> And it looks like there is a bug with the line counts within comments.

There is a misunderstanding. I used comment lines in example code in order to show the situation better. But this bug is not only related with comments. Lets say we have code like that:

```css
.selector {
  background-color: #fff;
}




 #Eighth-line-this-line-is-not-empty-again {
   content: "isee";
 }
```

Stylelint says:

```shell
 7:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
 8:1  ×  Expected no more than 2 empty line(s)   max-empty-lines
```

**Comment by jeddy3** [comment]
@Leucetios Thanks for the clarification.

**Comment by GorwinJororis** [comment]
Well, after one year, I assumed noone else tried to fix this. Also you have added `ignore: ['comments']` to `max-empty-lines` which is good thing. But this issue has to be fixed because rule is not working way it should be, as can as I remember, even tests for this rule is wrong @evilebottnawi @davidtheclark @jeddy3

**Comment by evilebottnawi** [comment]
@GorwinJororis it is very deep problem. Problem in `strigify` function of parser (we use this to get source code) and postcss return comments in `/* Comment */` format, but original code contain `// Comment`, some problem with other nodes exists. You can try to solve this, but it is not easy.

**Comment by GorwinJororis** [comment]
@evilebottnawi yes issue is exactly what you said. I can try again but idk if I can accomplish this or not.
